source/FormationPattern.cpp

```diff
@@ -93,7 +205,228 @@ void FormationPattern::SetName(const std::string &name)
+FormationPattern::PositionIterator FormationPattern::begin(
+	double diameterToPx, double widthToPx, double heightToPx,
+	double centerBodyRadius, unsigned int shipsToPlace) const
+{
+	return FormationPattern::PositionIterator(*this, diameterToPx, widthToPx,
+		heightToPx, centerBodyRadius, shipsToPlace);
+}
+
+
+
+// Get the number of lines (and arcs) in this formation.
+unsigned int FormationPattern::Lines() const
+{
+	return lines.size();
+}
+
+
+
+// Get the number of repeat sections for the given arc or line.
+unsigned int FormationPattern::Repeats(unsigned int lineNr) const
+{
+	if(lineNr >= lines.size())
+		return 0;
+	return lines[lineNr].repeats.size();
+}
+
+
+
+// Get the number of positions on an arc or line.
+unsigned int FormationPattern::Positions(unsigned int ring, unsigned int lineNr, unsigned int repeatNr) const
+{
+	// Retrieve the relevant line.
+	if(lineNr >= lines.size())
+		return 0;
+	const Line &line = lines[lineNr];
+
+	int lineRepeatPositions = line.positions;
+
+	// For the very first ring, only the initial positions are relevant.
+	if(ring > 0)
+	{
+		// For later rings we need to have repeat sections to perform repeating.
+		if(repeatNr >= line.repeats.size())
+			return 0;
+
+		lineRepeatPositions += line.repeats[repeatNr].repeatPositions * ring;
+	}
+
+	// If we skip positions, then remove them from the counting.
+	if(line.skipFirst && lineRepeatPositions > 0)
+		--lineRepeatPositions;
+	if(line.skipLast && lineRepeatPositions > 0)
+		--lineRepeatPositions;
+
+	// If we are in a later ring, then skip lines that don't repeat.
+	if(lineRepeatPositions < 0)
+		return 0;
+
+	return lineRepeatPositions;
+}
+
+
+
+// Check if a certain arc or line is centered.
+bool FormationPattern::IsCentered(unsigned int lineNr) const
+{
+	// Returns false if we have an invalid lineNr or when the line is not centered.
+	return lineNr < lines.size() && lines[lineNr].centered;
+}
+
+
+
+// Get a formation position based on ring, line(or arc)-number and position on the line.
+Point FormationPattern::Position(unsigned int ring, unsigned int lineNr, unsigned int repeatNr,
+	unsigned int linePosition, double diameterToPx, double widthToPx, double heightToPx) const
+{
+	// First check if the inputs result in a valid line or arc position.
+	if(lineNr >= lines.size())
+		return Point();
+	const Line &line = lines[lineNr];
+	if(ring > 0 && repeatNr >= line.repeats.size())
+		return Point();
+
+	// Perform common start and end/anchor position calculations in pixels.
+	Point startPx = line.start.GetPx(diameterToPx, widthToPx, heightToPx);
+	Point endOrAnchorPx = line.endOrAnchor.GetPx(diameterToPx, widthToPx, heightToPx);
+
+	// Get the number of positions for this line or arc.
+	int positions = line.positions;
+
+	// Check if we have a valid repeat section and apply it to the common calculations if we have it.
+	const LineRepeat *repeat = nullptr;
+	if(ring > 0 && repeatNr < line.repeats.size())
```

Comment:
```suggestion
	if(ring > 0)
```
we've already checked that at the beginning of the function